x/mint/keeper/abci.go

```diff
@@ -37,19 +37,42 @@ func (k Keeper) BeginBlocker(ctx context.Context, ic types.InflationCalculationF
 		return err
 	}
 
+	// update minter's inflation and annual provisions
 	minter.Inflation = ic(ctx, minter, params, bondedRatio)
 	minter.AnnualProvisions = minter.NextAnnualProvisions(params, totalStakingSupply)
 	if err = k.Minter.Set(ctx, minter); err != nil {
 		return err
 	}
 
-	// mint coins, update supply
+	// calculate minted coins
 	mintedCoin := minter.BlockProvision(params)
 	mintedCoins := sdk.NewCoins(mintedCoin)
 
-	err = k.MintCoins(ctx, mintedCoins)
-	if err != nil {
-		return err
+	maxSupply := params.MaxSupply
+	// if maxSupply is not infinite, check against max_supply parameter
+	if !maxSupply.IsZero() {
+		if totalStakingSupply.Add(mintedCoins.AmountOf(params.MintDenom)).GT(maxSupply) {
+			// calculate the difference between maxSupply and totalStakingSupply
+			diff := maxSupply.Sub(totalStakingSupply)
```

Comment:
We should also consider the burned tokens, correct?

Reply:
now that we use total supply from bank, supply is updated every time right?

Reply:
Supply doesn't track the burned tokens I guess. If it's not handled, we should track them separately?

Reply:
but how do we get burned tokens ?

Reply:
@tac0turtle @alexanderbez ?

Reply:
That's the issue right. Let's say if the max limit is m, x tokens are already burned, the new max-tokens-limit should be `m-x` since `x` were minted some of time in the history. Otherwise burning tokens doesn't make sense in this case

Reply:
I always thought of it differently. That if you reduce the token then it gives you more to mint unless you reduce the total supply. I'm not sure how we will track this otherwise and on existing chains. Is there a need for this?

Reply:
We can just have 0 address for burning tokens. That way it will not be reduced from Total supply directly but we can have tally of burned tokens. For existing chains, if they are maintaining the data of burned tokens somewhere, they can just include in the upgrade handler to mint those tokens for 0 address. We need to change the burncoins functionality to send it to this 0 address

Reply:
Hmm I tend to disagree this is an overall change. It's worth talking to users first instead of making the change blindly.